### PR TITLE
feat: implement solid wrapper for web

### PR DESCRIPTION
**Pull request by moonlitgrace**
## Description

Implementation of Solid wrapper around the dotlottie-web library.

## Type of change

- [ ]  Patch: Bug (non-breaking change which fixes an issue)
- [X]  Minor: New feature (non-breaking change which adds functionality)
- [ ]  Major: Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] Update README with relative links
- [x] Fix husky

### CODE REVIEW

**Comment by moonlitgrace** [comment]
These coverage tests are failing (no changes to test files).
Screenshot 2024-06-07 at 00-42-38 Vitest Browser Runner

File: packages/solid/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+/**
+ * Copyright 2024 Design Barn Inc.
```
Comment (baseplate-admin) [inline packages/solid/src/index.ts:2]: > Copyright 2024 Design Barn Inc.


This notice can be removed

File: packages/solid/README.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+You can find the list of methods that can be used to control the playback of the animation [here](https://github.com/LottieFiles/dotlottie-web/blob/main/packages/web/README.md#methods).
+
+## Listening to Events
+
+`DotLottieSolid` component can receive a `dotLottieRefCallback` prop that can be used to get a reference to the [dotLottie](https://github.com/LottieFiles/dotlottie-web/blob/main/packages/web/README.md#apis) web player instance. This reference can be used to listen to player events emitted by the [dotLottie](https://github.com/LottieFiles/dotlottie-web/blob/main/packages/web/README.md#events) web instance.
+
+Here is an example:
+
+```tsx
+import Solid from 'solidjs';
```
Comment (baseplate-admin) [inline packages/solid/README.md:137]: > import Solid from 'solidjs';

I think the better example here is to import {createEffect,onCleanup} from `solid-js`

File: packages/solid/dev/App.tsx [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    dotLottie()?.addEventListener('pause', console.log);
+    dotLottie()?.addEventListener('stop', console.log);
+  });
+
+  onCleanup(() => {
+    dotLottie()?.removeEventListener('play', console.log);
+    dotLottie()?.removeEventListener('freeze', console.log);
+    dotLottie()?.removeEventListener('unfreeze', console.log);
+    dotLottie()?.addEventListener('pause', console.log);
+    dotLottie()?.addEventListener('stop', console.log);
```
Comment (baseplate-admin) [inline packages/solid/dev/App.tsx:80]: This can be simplified with a guard clause check

File: packages/solid/dev/App.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  function handleAutoResizeCanvas(): void {
+    setAutoResizeCanvas((prev) => !prev);
+  }
+
+  createEffect(() => {
+    dotLottie()?.addEventListener('play', console.log);
+    dotLottie()?.addEventListener('freeze', console.log);
+    dotLottie()?.addEventListener('unfreeze', console.log);
+    dotLottie()?.addEventListener('pause', console.log);
+    dotLottie()?.addEventListener('stop', console.log);
```
Comment (baseplate-admin) [inline packages/solid/dev/App.tsx:72]: This can be simplified with a guard clause check

File: packages/solid/dev/App.tsx [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+  function handlePlayOnHover(): void {
+    setPlayOnHover((prev) => !prev);
+  }
+
+  function handleAutoResizeCanvas(): void {
+    setAutoResizeCanvas((prev) => !prev);
+  }
+
+  createEffect(() => {
```
Comment (baseplate-admin) [inline packages/solid/dev/App.tsx:67]: Should this not be a [`onMount`](https://www.solidjs.com/tutorial/lifecycles_onmount) ?

**Review by baseplate-admin — CHANGES_REQUESTED** [review]
Awesome work.

Minor nitpicks

File: packages/solid/src/index.ts (hunk 1 above)
Comment (moonlitgrace) [reply packages/solid/src/index.ts:2]: Seems like dotlottie is using some kind of eslint plugin
https://www.npmjs.com/package/eslint-plugin-header

File: packages/solid/dev/App.tsx (hunk 5 above)
Comment (moonlitgrace) [reply packages/solid/dev/App.tsx:67]: For better debugging `createEffect` is more suitable in this case.
`dotLottie()` instance gets updated in each `play, pause` like method calls.

File: packages/solid/src/index.ts (hunk 1 above)
Comment (theashraf) [reply packages/solid/src/index.ts:2]: I think we should disable this rule from the eslint config file, since it's an open-source project under the MIT license

File: packages/solid/dev/App.tsx (hunk 3 above)
Comment (moonlitgrace) [reply packages/solid/dev/App.tsx]: Resolved on https://github.com/LottieFiles/dotlottie-web/pull/234/commits/1c2353d815dc7b715f1837c645527d928698679a

File: packages/solid/dev/App.tsx (hunk 4 above)
Comment (moonlitgrace) [reply packages/solid/dev/App.tsx]: Resolved on https://github.com/LottieFiles/dotlottie-web/pull/234/commits/1c2353d815dc7b715f1837c645527d928698679a

File: packages/solid/src/index.ts (hunk 1 above)
Comment (moonlitgrace) [reply packages/solid/src/index.ts]: Resolved on https://github.com/LottieFiles/dotlottie-web/pull/234/commits/3b153a4f532a77713bcf3676eeb510f4b6705501

**Comment by moonlitgrace** [comment]
Hi @theashraf, porting is almost done.
but these tests are failing, do you see anything odd? do I need to do anything to configure new package?

File: packages/solid/README.md (hunk 2 above)
Comment (moonlitgrace) [reply packages/solid/README.md]: Resolved on https://github.com/LottieFiles/dotlottie-web/pull/234/commits/a8ec1a3ea7d3210762a26d225c95a167fdb4d366

File: packages/solid/src/use-dotlottie.tsx [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+  });
+  // data
+  createEffect(() => {
+    const dotLottieInstance = dotLottie();
+
+    if (!dotLottieInstance) return;
+
+    if (typeof config.data == 'string' || config.data instanceof ArrayBuffer) {
+      dotLottieInstance.load({
+        data: config.data,
```
Comment (theashraf) [inline packages/solid/src/use-dotlottie.tsx:255]: On any config change, wouldn't that trigger the effect? I keep seeing in the example that play and load events are being triggered on changes that are not related to loading new data, such as changing the speed or frame interpolation.

I think we need to ensure that we're only triggering the effect for the specific property we want to react to, not for the entire configuration.

**Review by theashraf — COMMENTED** [review]
Thanks, @moonlitgrace, for your contribution. Overall, it looks great. I've just left a few comments.

Regarding the failing tests, please ensure you're using pnpm v8 and reinstall the dependencies. I've noticed that the lock file is using pnpm v9, which might be causing dependency-related issues or version conflicts.

File: .changeset/quick-berries-remain.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+---
+'@lottiefiles/dotlottie-solid': patch
+---
+
+Implement Solid wrapper for dotlottie-web library
```
[the remark this answers is not in the document]
Comment (moonlitgrace) [reply .changeset/quick-berries-remain.md]: Resolved on https://github.com/LottieFiles/dotlottie-web/pull/234/commits/87cc0b431c0524ae5d1bc84d1de81c01542958b9

File: packages/solid/package.json [hunk 8]
```diff
@@ -0,0 +1,74 @@
+{
+  "name": "@lottiefiles/dotlottie-solid",
+  "version": "0.0.1",
```
[the remark this answers is not in the document]
Comment (moonlitgrace) [reply packages/solid/package.json]: Resolved on https://github.com/LottieFiles/dotlottie-web/pull/234/commits/87cc0b431c0524ae5d1bc84d1de81c01542958b9

**Comment by moonlitgrace** [comment]
> Regarding the failing tests, please ensure you're using pnpm v8 and reinstall the dependencies. I've noticed that the lock file is using pnpm v9, which might be causing dependency-related issues or version conflicts.

Yep, I forgot to enable corepack.
Seems like now its fixed. cheers!

File: packages/solid/dev/App.tsx [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+];
+
+const App: Component = () => {
+  const [dotLottie, setDotlottie] = createSignal<DotLottie>();
+  const [loop, setLoop] = createSignal(true);
+  const [speed, setSpeed] = createSignal(1);
+  const [srcIdx, setSrcIdx] = createSignal(0);
+  const [useFrameInterpolation, setUseFrameInterpolation] = createSignal(false);
+  const [playOnHover, setPlayOnHover] = createSignal(false);
+  const [autoResizeCanvas, setAutoResizeCanvas] = createSignal(true);
```
[the remark this answers is not in the document]
Comment (moonlitgrace) [reply packages/solid/dev/App.tsx]: Resolved on https://github.com/LottieFiles/dotlottie-web/pull/234/commits/179a084c8e4e108697fbe7aa626f06e24104b6a8

File: packages/solid/src/use-dotlottie.tsx (hunk 6 above)
Comment (moonlitgrace) [reply packages/solid/src/use-dotlottie.tsx]: Yes good catch, that was cause of de-structuring reactive props, `...config` in specific effect.
resolved on https://github.com/LottieFiles/dotlottie-web/pull/234/commits/179a084c8e4e108697fbe7aa626f06e24104b6a8

File: packages/solid/package.json [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    "debounce": "^2.0.0"
+  },
+  "devDependencies": {
+    "cross-env": "7.0.3",
+    "solid-js": "^1.8.17",
+    "tsup": "^8.0.2",
+    "tsup-preset-solid": "^2.2.0",
+    "typescript": "^5.4.5",
+    "vite": "^5.2.12",
+    "vite-plugin-solid": "^2.10.2"
```
Comment (theashraf) [inline packages/solid/package.json:69]: I don't think we need these vite related dependencies anymore

File: packages/solid/package.json [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    "player",
+    "animation",
+    "web",
+    "canvas",
+    "javascript",
+    "thorvg"
+  ],
+  "scripts": {
+    "build": "tsup",
+    "dev": "vite serve dev",
```
Comment (theashraf) [inline packages/solid/package.json:49]: `tsup --watch`

File: packages/solid/package.json (hunk 10 above)
Comment (moonlitgrace) [reply packages/solid/package.json]: Resolved on https://github.com/LottieFiles/dotlottie-web/pull/234/commits/79677f98bb727a8ee983f9c04c12b6267943593d

File: packages/solid/package.json (hunk 11 above)
Comment (moonlitgrace) [reply packages/solid/package.json]: Resolved on https://github.com/LottieFiles/dotlottie-web/pull/234/commits/79677f98bb727a8ee983f9c04c12b6267943593d
